packages/aws-cdk/lib/api/garbage-collection/garbage-collector.ts

```diff
@@ -0,0 +1,389 @@
+    const numObjects = await this.numObjectsInBucket(s3, bucket);
+    const printer = new ProgressPrinter(numObjects, 1000);
+
+    debug(`Found bootstrap bucket ${bucket}`);
+
+    try {
+      const batches = 1;
+      const batchSize = 1000;
+      const currentTime = Date.now();
+      const graceDays = this.props.rollbackBufferDays;
+
+      debug(`Parsing through ${numObjects} objects in batches`);
+
+      // Process objects in batches of 1000
+      // This is the batch limit of s3.DeleteObject and we intend to optimize for the "worst case" scenario
+      // where gc is run for the first time on a long-standing bucket where ~100% of objects are isolated.
+      for await (const batch of this.readBucketInBatches(s3, bucket, batchSize, currentTime)) {
+        print(chalk.green(`Processing batch ${batches} of ${Math.floor(numObjects / batchSize) + 1}`));
+        printer.start();
+
+        const { isolated, notIsolated } = batch.reduce<{ isolated: S3Asset[]; notIsolated: S3Asset[] }>((acc, obj) => {
+          if (!activeAssets.contains(obj.fileName())) {
+            acc.isolated.push(obj);
+          } else {
+            acc.notIsolated.push(obj);
+          }
+          return acc;
+        }, { isolated: [], notIsolated: [] });
+
+        debug(`${isolated.length} isolated assets`);
+
+        let deletables: S3Asset[] = isolated;
+        let taggables: S3Asset[] = [];
+
+        if (graceDays > 0) {
+          debug('Filtering out assets that are not old enough to delete');
+          await this.parallelReadAllTags(s3, isolated);
+          // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+          deletables = await Promise.all(isolated.map(async (obj) => {
+            const shouldDelete = await obj.isolatedTagBefore(s3, new Date(currentTime - (graceDays * DAY)));
+            return shouldDelete ? obj : null;
+          })).then(results => results.filter((obj): obj is S3Asset => obj !== null));
+
+          // eslint-disable-next-line @cdklabs/promiseall-no-unbounded-parallelism
+          taggables = await Promise.all(isolated.map(async (obj) => {
+            const shouldTag = await obj.noIsolatedTag(s3);
+            return shouldTag ? obj : null;
+          })).then(results => results.filter((obj): obj is S3Asset => obj !== null));
+        }
+
+        debug(`${deletables.length} deletable assets`);
+        debug(`${taggables.length} taggable assets`);
+
+        if (this.permissionToDelete && deletables.length > 0) {
+          await this.parallelDelete(s3, bucket, deletables, printer);
+        }
+
+        if (this.permissionToTag && taggables.length > 0) {
+          await this.parallelTag(s3, bucket, taggables, currentTime, printer);
+        }
+
+        printer.reportScannedObjects(batch.length);
+
+        // We untag objects that are referenced in ActiveAssets and currently have the Isolated Tag.
+        let untaggables: S3Asset[] = await Promise.all(notIsolated.map(async (obj) => {
```

Comment:
Untagging feels tacked on. I'd rather you follow the order that's already there:

* Calculate all actions
* Perform all actions
* Print progress

In that order.